src/bio/bio_monitor.c

```diff
@@ -0,0 +1,541 @@
+		dev_health->bhm_inflights--;
+	}
+
+out:
+	/* Free I/O request in the competion callback */
+	spdk_bdev_free_io(bdev_io);
+}
+
+/* Get the SPDK device health state log and print all useful stats */
+void
+bio_bs_monitor(struct bio_xs_context *ctxt, uint64_t now)
+{
+	struct bio_health_monitoring	*dev_health;
+	struct spdk_bdev		*bdev;
+	struct spdk_nvme_cmd		 cmd;
+	int				 rc;
+	uint32_t			 numd, numdl, numdu;
+	uint32_t			 health_page_sz;
+
+	D_ASSERT(ctxt != NULL);
+	D_ASSERT(ctxt->bxc_blobstore != NULL);
+	dev_health = &ctxt->bxc_blobstore->bb_dev_health;
+	D_ASSERT(dev_health->bhm_io_channel != NULL);
+	D_ASSERT(dev_health->bhm_desc != NULL);
+
+	/*
+	 * TODO Decide on an appropriate period to query device health
+	 * stats. Currently set at 1 min.
+	 */
+	if (dev_health->bhm_stat_age + DAOS_SPDK_STATS_PERIOD >= now)
+		return;
+	dev_health->bhm_stat_age = now;
+
+	bdev = spdk_bdev_desc_get_bdev(dev_health->bhm_desc);
+	D_ASSERT(bdev != NULL);
+
+	/* Return if non-NVMe device */
+	if (get_bdev_type(bdev) != BDEV_CLASS_NVME)
+		return;
+
+	if (!spdk_bdev_io_type_supported(bdev, SPDK_BDEV_IO_TYPE_NVME_ADMIN)) {
+		D_ERROR("Bdev NVMe admin passthru not supported!\n");
+		return;
+	}
+
+	/* Check to avoid parallel SPDK device health query calls */
+	if (dev_health->bhm_inflights)
+		return;
+	dev_health->bhm_inflights++;
+
+	/* Prep NVMe command to get SPDK device health data */
+	health_page_sz = sizeof(struct spdk_nvme_health_information_page);
+	numd = health_page_sz / sizeof(uint32_t) - 1u;
+	numdl = numd & 0xFFFFu;
+	numdu = (numd >> 16) & 0xFFFFu;
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.opc = SPDK_NVME_OPC_GET_LOG_PAGE;
+	cmd.nsid = SPDK_NVME_GLOBAL_NS_TAG;
+	cmd.cdw10 = numdl << 16;
+	cmd.cdw10 |= SPDK_NVME_LOG_HEALTH_INFORMATION;
+	cmd.cdw11 = numdu;
+
+	/*
+	 * Submit an NVMe Admin command to get device health log page
+	 * to the bdev.
+	 */
+	rc = spdk_bdev_nvme_admin_passthru(dev_health->bhm_desc,
+					   dev_health->bhm_io_channel,
+					   &cmd,
+					   dev_health->bhm_health_buf,
+					   health_page_sz,
+					   get_spdk_log_page_completion,
+					   dev_health);
+	if (rc) {
+		D_ERROR("NVMe admin passthru (health log), rc:%d\n", rc);
+		dev_health->bhm_inflights--;
+	}
+}
+
+/* Print the io stat every few seconds, for debug only */
+void
+bio_xs_io_stat(struct bio_xs_context *ctxt, uint64_t now)
+{
+	struct spdk_bdev_io_stat	 stat;
+	struct spdk_bdev		*bdev;
+	struct spdk_io_channel		*channel;
+
+	/* check if IO_STAT_PERIOD environment variable is set */
+	if (io_stat_period == 0)
+		return;
+
+	if (ctxt->bxc_io_stat_age + io_stat_period >= now)
+		return;
+
+	if (ctxt->bxc_desc != NULL) {
+		channel = spdk_bdev_get_io_channel(ctxt->bxc_desc);
+		D_ASSERT(channel != NULL);
+		spdk_bdev_get_io_stat(NULL, channel, &stat);
+		spdk_put_io_channel(channel);
+
+		bdev = spdk_bdev_desc_get_bdev(ctxt->bxc_desc);
+		D_ASSERT(bdev != NULL);
+
+		D_PRINT("SPDK IO STAT: xs_id[%d] dev[%s] read_bytes["DF_U64"], "
```

Comment:
Not immediately, but we must look to print this with different units than bytes, just harder to comprehend, I know this is for the developers, still its useful